src/crypto.rs

```diff
@@ -12,66 +14,110 @@ use self::rustc_serialize::base64::{self, FromBase64, ToBase64};
-    fn from(err: &'static str) -> CryptoError {
-        CryptoError::Custom(err)
+impl Error for CryptoError {
+    fn description(&self) -> &str {
+        match *self {
+            CryptoError::Custom(ref string) => string,
+            CryptoError::Io(ref err) => err.description(),
+        }
     }
 }
 
-impl From<std::io::Error> for CryptoError {
-    fn from(err: std::io::Error) -> CryptoError {
-        CryptoError::Io(err)
+impl Display for CryptoError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self{
+            CryptoError::Custom(ref string) => string.fmt(f),
+            CryptoError::Io(ref err) => write!(f, "IO error: {}", err),
+        }
     }
 }
 
-impl From<openssl::ssl::error::SslError> for CryptoError {
-    fn from(err: openssl::ssl::error::SslError) -> CryptoError {
-        CryptoError::Ssl(err)
+macro_rules! from_error {
+    ( $orig:ty, $enum_type:ident ) => {
+        impl From<$orig> for CryptoError {
+            fn from(err: $orig) -> CryptoError {
+                CryptoError::$enum_type(err)
+            }
+        }
     }
 }
 
+from_error!(String, Custom);
+from_error!(std::io::Error, Io);
+
 
 /// A named key pair, for use in JWS signing.
 #[derive(Clone)]
 pub struct NamedKey {
     id: String,
     key: PKey,
+    valid_from: Option<DateTime<UTC>>,
 }
 
 
 impl NamedKey {
     /// Creates a NamedKey by reading a `file` path and generating an `id`.
     pub fn from_file(filename: &str) -> Result<NamedKey, CryptoError> {
         let mut file = File::open(filename)?;
-        let mut file_contents = String::new();
-        file.read_to_string(&mut file_contents)?;
+        let mut pem = String::new();
+        file.read_to_string(&mut pem)?;
+
+        let pkey = PKey::private_key_from_pem(&mut pem.as_bytes())
+            .map_err(|err| format!("failed to parse key '{}': {}", filename, err))?;
+
+        let meta = Self::meta_from_pem(&pem)
+            .map_err(|err| format!("failed to parse key '{}': {}", filename, err))?;
 
-        NamedKey::from_pem_str(&file_contents)
+        let mut key = NamedKey::from_pkey(pkey)?;
+        key.valid_from = meta.valid_from;
+        Ok(key)
     }
 
     /// Creates a NamedKey from a PEM-encoded str.
     pub fn from_pem_str(pem: &str) -> Result<NamedKey, CryptoError> {
-        let pkey = PKey::private_key_from_pem(&mut pem.as_bytes())?;
+        let pkey = PKey::private_key_from_pem(&mut pem.as_bytes())
+            .map_err(|err| format!("failed to parse key: {}", err))?;
 
-        NamedKey::from_pkey(pkey)
+        let meta = Self::meta_from_pem(&pem)
+            .map_err(|err| format!("failed to parse key: {}", err))?;
+
+        let mut key = NamedKey::from_pkey(pkey)?;
+        key.valid_from = meta.valid_from;
+        Ok(key)
+    }
+
+    /// Parses metadata from a PEM file.
+    fn meta_from_pem(pem: &str) -> Result<PrivateKeyMetadata, String> {
+        toml::decode_str(
+            &pem.lines()
```

Comment:
I didn't look, actually, but also think what this is doing is fairly specific. Most libs will only look for the key block.